### PR TITLE
Extract heap profiles from the event log

### PR DESCRIPTION
**Pull request by mpickering**
cc @bgamari @osa1

### CODE REVIEW

**Review by jaspervdj — APPROVED** [review]

**Comment by jaspervdj** [comment]
Looks good to me -- should I go ahead and merge this @mpickering or do you want to wait for comments from Ben or Ömer?

**Comment by mpickering** [comment]
I just need to wait to see if @bgamari is OK with being listed as a mentor. Should I put my email down as well?

**Review by gwils — APPROVED** [review]

File: content/ideas/eventlog-profiling.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+---
+title: Extract heap profiles from the event log
+---
+
+The [event log](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/eventlog-formats.html#eventlog-encodings) is a general purpose mechanism for understanding what a
+Haskell program is doing. It is a binary stream of low-level information logging
+what the RTS is doing.
+
+Ben Gamari [recently extended](https://ghc.haskell.org/trac/ghc/ticket/11094) the format so that heap profiling events are also
+written to the log. This means that, in theory, all the information present in a
```
Comment (maoe) [inline content/ideas/eventlog-profiling.md]: IIRC the heap profiling in the eventlog framework isn't complete. Last I checked at least retainer and biographical profiling was not supported yet. Could we add it to the list of goals?

File: content/ideas/eventlog-profiling.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+reasons. Firstly it means that the specific
+logic relating to making `.hp` files can be removed from the compiler. Secondly,
+users can better correlate over events in their program with the program's memory
+usage. This would [implement a feature](https://ghc.haskell.org/trac/ghc/ticket/3021) present in `nhc98` from over 20 years ago.
+
+The goal of this project is to eliminate the this redundancy by making the event log
+the primary way to produce heap profiles.
+
+1. Understand the event log and heap profile format. Write a program to convert
+    the event log to a heap profile.
```
Comment (maoe) [inline content/ideas/eventlog-profiling.md:27]: Why is this step necessary? It'd be unnecessary with the following step, no?

File: content/ideas/eventlog-profiling.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+2. Modify tools such as `hp2pretty` to directly consume an event log.
+3. Modify the RTS so that turning on the heap profiler just means emitting the event log.
+4. Add support for marking user events on a heap profile.
```
Comment (maoe) [inline content/ideas/eventlog-profiling.md]: Could you elaborate on this? FWIW there will be [`traceBinaryEvent`](https://phabricator.haskell.org/D5007) available in GHC 8.8 that supports emitting user defined messages.

**Review by maoe — COMMENTED** [review]
I added some comments and questions.

File: content/ideas/eventlog-profiling.md (hunk 1 above)
Comment (mpickering) [reply content/ideas/eventlog-profiling.md]: Good point. I'll add that.

File: content/ideas/eventlog-profiling.md (hunk 2 above)
Comment (mpickering) [reply content/ideas/eventlog-profiling.md:27]: Yes, however `1.` can be achieved without having to get to grips with the code of `hp2pretty` at all. I wanted to try to break down the project into as small chunks as possible.

File: content/ideas/eventlog-profiling.md (hunk 3 above)
Comment (mpickering) [reply content/ideas/eventlog-profiling.md]: Yes `traceEvent` and `traceBinaryEvent` are the two functions I'm thinking of. The point of this goal is to display this information on the heap profile graph.

image

**Comment by mpickering** [comment]
Ben said it was fine to list him as a mentor but that I should act as the primary point of contact.

File: content/ideas/eventlog-profiling.md (hunk 1 above)
Comment (maoe) [reply content/ideas/eventlog-profiling.md]: Thanks

File: content/ideas/eventlog-profiling.md (hunk 2 above)
Comment (maoe) [reply content/ideas/eventlog-profiling.md:27]: I see. That makes sense.

File: content/ideas/eventlog-profiling.md (hunk 3 above)
Comment (maoe) [reply content/ideas/eventlog-profiling.md]: Thanks for the clarification. Sounds good to me.

**Review by maoe — APPROVED** [review]
This proposal looks good to me.

**Pull request merged**